zealot/fetcher/fetcher.ts

```diff
@@ -1,72 +1,108 @@
 import {url} from "../util/utils"
 import {parseContentType} from "./contentType"
-import {Enhancer, ZealotUploadPayload, ZResponse} from "../types"
+import {Enhancer, ZResponse} from "../types"
 import {createIterator} from "./iterator"
 import {createStream} from "./stream"
 import {createError} from "../util/error"
-import {createPushableIterator} from "./pushable-iterator"
-import {parseLines} from "../ndjson/lines"
+import {EventSourcePolyfill} from "event-source-polyfill"
+import nodeFetch from "node-fetch"
+import stream from "stream"
+import {AbortController as NodeAbortController} from "node-abort-controller"
 
 export type FetchArgs = {
   path: string
-  method: string
-  body?: string | FormData | ReadableStream
+  method?: string
+  body?: string | FormData | ReadableStream | NodeJS.ReadableStream
   headers?: Headers
   enhancers?: Enhancer[]
   signal?: AbortSignal
+  useNodeFetch?: boolean
 }
 
-export function createFetcher(host: string) {
+const fetchWithTimeout = async (
+  baseUrl: string,
+  args: FetchArgs
+): Promise<Response> => {
+  const {path, method, body, signal, headers, useNodeFetch} = args
+  const [wrappedSignal, clearTimeout] = useTimeoutSignal(signal)
+  if (body instanceof stream.Readable) {
+    body.once("data", () => clearTimeout())
+    body.once("start", () => clearTimeout())
```

Comment:
This additional "start" listener is kind of a bummer, but was necessary for the brimcap stream pipeline where the stdout stream's data can take some time to begin to produce data, while the stderr stream's data progress gives us a quicker status report. Clearing the same timeout more than once has no realistic consequence.